### ISSUE TITLE
XLNet / sentence padding

### ISSUE THREAD

**Issue by cherepanovic** [huggingface/transformers#1013]
My samples have different lengths and I want to apply the padding to bring them to the same length, because my purpose is to create sentence embeddings batchwise. For that all sentences must have the same length, otherwise it is not possible to create a tensor. 

How does padding work in use of XLNet model? 

the snippet below shows my first try to do it with XLNet, I apply maxpooling on the model output. 

```
class MaxPoolingChannel(torch.nn.AdaptiveMaxPool1d):
    def forward(self, input):
        input = input[0]
        input = input.transpose(1,2)
        result =  torch.nn.AdaptiveMaxPool1d(1)(input)
        return result.transpose(2,1)

tokenizer = XLNetTokenizer.from_pretrained('xlnet-base-cased')
model = XLNetModel.from_pretrained('xlnet-base-cased')
model.eval()

input_ids = torch.tensor(tokenizer.encode("Hello, my dog is cute")).unsqueeze(0)  # Batch size 1
input_ids_2 = torch.tensor(tokenizer.encode("Hello, my dog is cute <pad>")).unsqueeze(0)
input_ids_3 = torch.tensor(tokenizer.encode("Hello, my dog is cute <pad> <pad>")).unsqueeze(0)


with torch.no_grad():
    model = torch.nn.Sequential(model, MaxPoolingChannel(1))
    res = model(input_ids)
    res_2 = model(input_ids_2)
    res_3 = model(input_ids_3)

print(cosine_similarity(res.detach().numpy()[0][0],res_2.detach().numpy()[0][0]))
print(cosine_similarity(res.detach().numpy()[0][0],res_3.detach().numpy()[0][0]))
```

There is a thread #790 (about document embeddings), however the point in regard to padding in XLNet has not been touched.  

Thanks!

**Comment by LysandreJik** [comment]
Hi!

By concatenating the `<pad>` value to the end of your sentences you are successfully padding them. It can be observed by identifying the encoded sentence, which shows that a `5` value (which is the padding index in the tokenizer dictionary) is appended to the end of your token sequences.

Once you have padded your sentences, you can tell the model to ignore the padded values by specifying an `attention_mask` or an `input_mask`, as described in [the documentation.](https://huggingface.co/pytorch-transformers/model_doc/xlnet.html#xlnetmodel)

**Comment by cherepanovic** [comment]
@LysandreJik 

I tried it out, however there are some distance between the vectors after having applied padding.

```
input_ids = torch.tensor(tokenizer.encode("Hello, my dog is cute")).unsqueeze(0)  # Batch size 1
input_ids_2 = torch.tensor(tokenizer.encode("Hello, my dog is cute <pad>")).unsqueeze(0)
perm_mask_2 = torch.zeros((1, input_ids_2.shape[1], input_ids_2.shape[1]), dtype=torch.float)
perm_mask_2[:, :, -1] = 1.0  # Previous tokens don't see last token
input_ids_3 = torch.tensor(tokenizer.encode("Hello, my dog is cute <pad> <pad>")).unsqueeze(0)
perm_mask_3 = torch.ones((1, input_ids_3.shape[1], input_ids_3.shape[1]), dtype=torch.float)
perm_mask_3[:, :, 0:-2] = 0.0  # Previous tokens don't see last token

with torch.no_grad():

    outputs = model(input_ids)
    res = MaxPoolingChannel(1)(outputs)
    outputs_2 = model(input_ids_2, perm_mask=perm_mask_2)
    res_2 = MaxPoolingChannel(1)(outputs_2)
    outputs_3 = model(input_ids_3, perm_mask=perm_mask_3)
    res_3 = MaxPoolingChannel(1)(outputs_3)

print(cosine_similarity(res.detach().numpy()[0][0],res.detach().numpy()[0][0]))   #0.9999999999999998
print(cosine_similarity(res.detach().numpy()[0][0],res_2.detach().numpy()[0][0])) #0.9986772544514477
print(cosine_similarity(res.detach().numpy()[0][0],res_3.detach().numpy()[0][0])) #0.997999551655906
```

What do I wrong? 

Thanks a lot!

**Comment by cherepanovic** [comment]
I tried it also with the next word prediction (different amount of `<pad>` in the sentence, target_mask and the perm_mask were adapted),`<pad>` affects the result (as distances were different)

```
# We show how to setup inputs to predict a next token using a bi-directional context.
input_ids = torch.tensor(tokenizer.encode("Hello, my dog is very <mask>")).unsqueeze(0)  # We will predict the masked token
perm_mask = torch.zeros((1, input_ids.shape[1], input_ids.shape[1]), dtype=torch.float)
perm_mask[:, :, -1] = 1.0  # Previous tokens don't see last token
target_mapping = torch.zeros((1, 1, input_ids.shape[1]), dtype=torch.float)  # Shape [1, 1, seq_length] => let's predict one token
target_mapping[0, 0, -1] = 1.0  # Our first (and only) prediction will be the last token of the sequence (the masked token)
outputs = xlnet_model(input_ids, perm_mask=perm_mask, target_mapping=target_mapping)
next_token_logits = outputs[0]  # Output has shape [target_mapping.size(0), target_mapping.size(1), config.vocab_size]
print(target_mapping)

# We show how to setup inputs to predict a next token using a bi-directional context.
input_ids = torch.tensor(tokenizer.encode("<pad> <pad> <pad> Hello, my dog is very  <mask> <pad> <pad> <pad>")).unsqueeze(0)  # We will predict the masked token
perm_mask = torch.ones((1, input_ids.shape[1], input_ids.shape[1]), dtype=torch.float)
perm_mask[:, :, 0:-4] = 0.0 
perm_mask[:, :, 0:3] = 1 
target_mapping = torch.zeros((1, 1, input_ids.shape[1]), dtype=torch.float)  # Shape [1, 1, seq_length] => let's predict one token
target_mapping[0, 0, -4] = 1.0  # Our first (and only) prediction will be the last token of the sequence (the masked token)

outputs = xlnet_model(input_ids, perm_mask=perm_mask, target_mapping=target_mapping)
next_token_logits2 = outputs[0]  # Output has shape [target_mapping.size(0), target_mapping.size(1), config.vocab_size]

distance = cosine_similarity(next_token_logits2.detach().numpy().reshape(-1,1)[0], next_token_logits.detach().numpy().reshape(-1,1)[0])
```

**Comment by cherepanovic** [comment]
@LysandreJik

I tried it out (both of keywords {attention_mask, input_mask}), however there are still some distance between the vectors after having applied padding.

```
input_ids = torch.tensor(tokenizer.encode("Hello, my dog is cute")).unsqueeze(0)  # Batch size 1
input_ids_2 = torch.tensor(tokenizer.encode("Hello, my dog is cute <pad>")).unsqueeze(0)
mask_2 = torch.zeros((1, input_ids_2.shape[1], input_ids_2.shape[1]), dtype=torch.float)
mask_2[:, :, -1] = 1.0  # Previous tokens don't see last token
input_ids_3 = torch.tensor(tokenizer.encode("Hello, my dog is cute <pad> <pad>")).unsqueeze(0)
mask_3 = torch.ones((1, input_ids_3.shape[1], input_ids_3.shape[1]), dtype=torch.float)
mask_3[:, :, 0:-2] = 0.0  # Previous tokens don't see last token


with torch.no_grad():

    outputs = model(input_ids)
    res = MaxPoolingChannel(1)(outputs)
    outputs_2 = model(input_ids_2, attention_mask=mask_2[:, 0])
    res_2 = MaxPoolingChannel(1)(outputs_2)
    outputs_3 = model(input_ids_3, attention_mask=mask_3[:, 0])
    res_3 = MaxPoolingChannel(1)(outputs_3)

print(cosine_similarity(res.detach().numpy()[0][0],res.detach().numpy()[0][0]))   #0.9999999999999998
print(cosine_similarity(res.detach().numpy()[0][0],res_2.detach().numpy()[0][0])) #0.9825270291286329
print(cosine_similarity(res.detach().numpy()[0][0],res_3.detach().numpy()[0][0])) #0.9825800158871332
```

What do I wrong?

Thanks a lot!

**Comment by cherepanovic** [comment]
I did a comparison in all dimensions between the outputs, they are different 

```
for i in range(outputs[0][0,:].shape[0]):
    print("outputs/outputs_3",cosine_similarity(outputs[0][0,i].numpy(),outputs_3[0][0,i].numpy()))

for i in range(outputs_2[0][0,:].shape[0]):
    print("outputs_2/outputs_3",cosine_similarity(outputs_2[0][0,i].numpy(),outputs_3[0][0,i].numpy()))


'''
outputs/outputs_3 0.9799852332583252
outputs/outputs_3 0.8951206569873185
outputs/outputs_3 0.8771695102750957
outputs/outputs_3 0.8325002575851723
outputs/outputs_3 0.9270082459388221
outputs/outputs_3 0.9484341771783724
outputs/outputs_3 0.8390104068562912
outputs_2/outputs_3 0.998496497302746
outputs_2/outputs_3 0.9918327763853063
outputs_2/outputs_3 0.9673674240861558
outputs_2/outputs_3 0.9379596336305814
outputs_2/outputs_3 0.9951713310140066
outputs_2/outputs_3 0.9949657485717505
outputs_2/outputs_3 0.9935681656710073
outputs_2/outputs_3 0.9999999669357101
'''
```

**Comment by cherepanovic** [comment]
the values are not the same, they are a slightly different (an exact consistency is not possible?)

```
-------------------
0 -0.7767028212547302
1 0.15364784002304077
2 -0.5269558429718018
3 -0.04860188066959381
4 0.14985302090644836
5 -0.6860541105270386
6 -1.598402738571167
-------------------
0 -0.7766993641853333
1 0.15364792943000793
2 -0.5269524455070496
3 -0.04859305918216705
4 0.1498618721961975
5 -0.6860424280166626
6 -1.5983952283859253
7 -0.921322226524353
8 -0.6499249935150146
```

would be the result of some picked dimension of an unpadded and a padded sentence. 

```
for i in range(outputs[0][0,:].shape[0]):
    print(i, outputs[0][0][i, 0].item())
print('-------------------')
for i in range(outputs_3[0][0,:].shape[0]):
    print(i, outputs_3[0][0][i, 0].item())
```

**Comment by cherepanovic** [comment]
the values are not the same, they are a slightly different (an exact consistency is not possible?)

```
-------------------
0 -0.7767028212547302
1 0.15364784002304077
2 -0.5269558429718018
3 -0.04860188066959381
4 0.14985302090644836
5 -0.6860541105270386
6 -1.598402738571167
-------------------
0 -0.7766993641853333
1 0.15364792943000793
2 -0.5269524455070496
3 -0.04859305918216705
4 0.1498618721961975
5 -0.6860424280166626
6 -1.5983952283859253
7 -0.921322226524353
8 -0.6499249935150146
```

would be the result of some picked dimension of an unpadded and a padded sentence. 

```
for i in range(outputs[0][0,:].shape[0]):
    print(i, outputs[0][0][i, 0].item())
print('-------------------')
for i in range(outputs_3[0][0,:].shape[0]):
    print(i, outputs_3[0][0][i, 0].item())
```

**Comment by Areshey** [comment]
> Hi!
> 
> By concatenating the `<pad>` value to the end of your sentences you are successfully padding them. It can be observed by identifying the encoded sentence, which shows that a `5` value (which is the padding index in the tokenizer dictionary) is appended to the end of your token sequences.
> 
> Once you have padded your sentences, you can tell the model to ignore the padded values by specifying an `attention_mask` or an `input_mask`, as described in [the documentation.](https://huggingface.co/pytorch-transformers/model_doc/xlnet.html#xlnetmodel)



> Hi!
> 
> By concatenating the `<pad>` value to the end of your sentences you are successfully padding them. It can be observed by identifying the encoded sentence, which shows that a `5` value (which is the padding index in the tokenizer dictionary) is appended to the end of your token sequences.
> 
> Once you have padded your sentences, you can tell the model to ignore the padded values by specifying an `attention_mask` or an `input_mask`, as described in [the documentation.](https://huggingface.co/pytorch-transformers/model_doc/xlnet.html#xlnetmodel)

I'd like to highlight that I encountered the same problem. the position IDs of padding tokens are set to 5 instead of 0. This value may either remain unchanged or undergo modifications in the future.
